### ISSUE TITLE
Dataset shuffling feature

### ISSUE THREAD

**Issue by beru** [tiny-dnn/tiny-dnn#562]
It looks like shuffling processing order of dataset is common but this feature is missing in this library.

**Comment by beru** [comment]
Right now, batch samples are copied in `network::train_onebatch` method.
Ideally, list of indexes should be used and huge amount of memory copy shouldn't be used.

**Comment by Randl** [comment]
Using [`std::shuffle`](http://www.cplusplus.com/reference/algorithm/shuffle/) with same seed seems to be simplest solution

**Comment by beru** [comment]
@Randl To use `std::shuffle` against input samples, we need to change type of inputs from
`std::vector<tensor_t>` to `std::vector<tensor_t*>` to avoid instance copy, alloc, dealloc.

In `network::train` method, `network::fit` method is called and in there, epoch and batch enumeration loops appear.

https://github.com/tiny-dnn/tiny-dnn/blob/33052fe26030a841fa729a8fa60744d0f46abee2/tiny_dnn/network.h#L761

Somewhat annoying part is, in the loop `network::train_once` is called and in there `network::train_onebatch` is called. and inside that method a lot of memory copy and allocation takes place.

```cpp
    std::copy(&in[0], &in[0] + batch_size, &in_batch_[0]);
    std::copy(&t[0], &t[0] + batch_size, &t_batch_[0]);
    std::vector<tensor_t> t_cost_batch =
      t_cost ? std::vector<tensor_t>(&t_cost[0], &t_cost[0] + batch_size)
             : std::vector<tensor_t>();
```
https://github.com/tiny-dnn/tiny-dnn/blob/33052fe26030a841fa729a8fa60744d0f46abee2/tiny_dnn/network.h#L832

**Comment by edgarriba** [comment]
Ideally the shuffling should be done before loading the data for each batch. However, there are some times that's not possible since you have all loaded in memory. In that case I think the best (and common) standard solution is having a proper DatasetLoader which takes care about batch generation, preprocessing and possibly parallelizing with queues. Please check [Pytorch](https://github.com/pytorch/pytorch/tree/master/torch/utils/data) solution for that.

In regards of random seed, there's already a discussion about it. https://github.com/tiny-dnn/tiny-dnn/issues/285

Possibly adding a small patch with `std::shuffle` is enough for now, since the Dataset loader might take some time and improve training it's not in the short term agenda.

**Comment by beru** [comment]
That modular design seems very appropriate. But it's very far.

I'd like to know how new Tensor class can prevent copying samples data when to ready a batch. It seems to me using a pointer (naked or smart) for each sample is required to avoid instance copy.

**Comment by edgarriba** [comment]
Didn't thought about that yet. Maybe through view?
